src/components/input/input.jsx

```diff
@@ -1,37 +1,91 @@
 import PropTypes from 'prop-types';
 import React from 'react';
+import cautionSvg from '../../svgs/icon-caution-fill.svg';
+import Icon from '../icon/icon.jsx';
 import styles from './input.css';
-import labelPropType from './label-prop-type';
 
-/**
- * Base input component. Accepts any prop that an input can take.
- * @see [RFC](https://github.com/mavenlink/rfc/pull/75)
- * @see [Design spec](https://zpl.io/295Yp7p)
- */
-export default function Input({ 'aria-labelledby': ariaLabelledBy, className, id, ...rest }) {
+function getLabelClassName(className, invalid) {
+  if (className) return className;
+  return invalid ? styles['invalid-label'] : styles.label;
+}
+
+function getClassName(className, invalid) {
+  if (className) return className;
+  return invalid ? styles['invalid-input'] : styles.input;
+}
+
+export default function Input(props) {
   return (
-    <input
-      {...rest}
-      aria-labelledby={ariaLabelledBy}
-      className={className}
-      id={id}
-    />
+    <React.Fragment>
+      <label htmlFor={props.id} className={getLabelClassName(props.cssLabel, props.invalid)}>
+        {props.label}
+      </label>
+      <div className={props.cssContainer}>
+        <input
+          className={getClassName(props.className, props.invalid)}
```

Comment:
I'm concerned about the number of configurations here. Are there things we could do, like advocate for the usage of native handlers for some of these?

Reply:
This is an opinionated implementation. 

As you can see from the code removed section, the MDS Input v1 was just that -- it passed through all props onto the `<input>` element. We thought that would be the best for MDS library but we saw incorrect usages of our MDS Input. Since we can just pass any kind of `type` into the MDS input, we ended up in a situation where a feature team used our MDS input as a checkbox -- which basically gutted the whole thing. 

Now, I think I understand the whole battle between configuration vs composition. While we do see many configuration options, I do not think this is a "configurable" component -- I think that is usually applied to components that change their behavior based on their properties. In this case, we are whitelisting the configuration options to the `<input>` -- which means we _want_ to control how much configuration is allowed for native elements.

Reply:
Time to have another look at typescript :) We're having same sorts of nuanced issues at my new place and have even gone so far as to throw a fatal Error if unrecognized props are passed in—this starts to make the code less readable and it's not fault tolerant per se. Typescript lets you know at compile time, you can just declare the interface you expect, and all is good. It IS a PITA to configure though.

Something to maybe consider longer term I  dunno :)

Reply:
I'm definitely not comfortable with the idea that implementation issues should be solved by just switching languages, there's a wealth of baggage for any language. 😄 I do like the opinionated approach, but perhaps we could forward native event handlers specifically? That way our interface maintains at least part of the underlying HTML native interface, and the part that we need for proper usage of the component while also supporting wrappers for more advanced behavior.

Reply:
There are no native event handles in React! React uses its own event system (and, sure, it ties it into native event handles but that is abstracted away from us React developers).

Reply:
@roblevintennis Hello! 👋 

Yeah, I'm on the same boat as always: I don't want to adopt yet another non-standard thing if I can post-pone it as long as possible. We adopted React and CSS Modules but we did wait a good 4 years before doing any of that. Hah, maybe in 4 years I'll consider TS. ;)

Reply:
I feel ya kinda ;) In this spirit of coding towards standards, I'd be very curious if esnext et al introduce some sort of notion of interfaces; although it's sort of useless as a runtime only thing imo (main thing is to get compile / transpile knowledge of bad inputs).

If you think about it, really, Webpack is a very specific way to set things up and many folks consider it extremely painful (folks at my new gig are in favor of using create react app or gatsby or something because they've tired of fighting Webpack). React leaves itself around in your code (e.g. read the philosophy of Svelte which compiles itself away). CSS Modules is really just a nifty spec which ala css-loader and Webpack does some name munging tricks, and as we know does not work perfectly (although I'm still using it as it's the least of all unfortunate options).

All to say, our jobs and lives in web-land are still a mess and not really so standards-based 😋 I am sad that my new gig is still quite rooted in Sass but maybe we'll get on the PostCSS train in 2020.

Reply:
> there's a wealth of baggage for any language

Typescript is not really a language but more of a superset of JavaScript. Yes, there's some baggage, but also some benefits. Especially if you like to use inversion of control (which I know you loathe from a particular meeting I recall where we brought up DI and you were a bigtime detractor).